File: src/class059/cpp/Code04_AlienDictionary.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// 火星词典
// 现有一种使用英语字母的火星语言
// 这门语言的字母顺序对你来说是未知的。
// 给你一个来自这种外星语言字典的字符串列表 words
// words 中的字符串已经 按这门新语言的字母顺序进行了排序 。
// 如果这种说法是错误的，并且给出的 words 不能对应任何字母的顺序，则返回 ""
// 否则，返回一个按新语言规则的 字典递增顺序 排序的独特字符串
// 如果有多个解决方案，则返回正常字典序最小的那个
// words中的单词一定都是小写英文字母组成的
// 测试链接 : https://www.lintcode.com/problem/892/
class Solution {
 public:
  string alienOrder(vector<string> &words) {
    vector<int> indegree(26, -1);
    for (string &w : words) {
      for (char c : w) {
        indegree[c - 'a'] = 0;
      }
    }

    vector<vector<int>> graph(26);

    for (int i = 0, j, len; i < words.size() - 1; ++i) {
      string &cur = words[i];
      string &next = words[i + 1];
      j = 0;
      len = min(cur.size(), next.size());

      for (; j < len; ++j) {
        if (cur[j] != next[j]) {
          graph[cur[j] - 'a'].push_back(next[j] - 'a');
          indegree[next[j] - 'a']++;
          break;
        }
      }

      if (j < cur.size() && j == next.size()) {
        return "";
      }
    }

    priority_queue<int, vector<int>, greater<>> queue;
    int kinds = 0;
    for (int i = 0; i < 26; ++i) {
      if (indegree[i] != -1) {
        kinds++;
      }

      if (indegree[i] == 0) {
        queue.push(i);
      }
    }

    string ans;
    while (!queue.empty()) {
      int cur = queue.top();
      queue.pop();
      ans.push_back(cur + 'a');

      for (int next : graph[cur]) {
        if (--indegree[next] == 0) {
          queue.push(next);
        }
      }
    }

    return ans.length() == kinds ? ans : "";
  }
};

auto __FAST_IO__ = []() noexcept -> int {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  return 0;
};

File: src/class059/java/Code04_AlienDictionary.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;

// 外星人字典
// 测试链接 : https://www.lintcode.com/problem/892/description?showListFe=true&page=1&submissionStatus=ACCEPTED&pageSize=50
public class Code04_AlienDictionary {
    public static String alienOrder(String[] words) {
        int[] indegree = new int[26];
        Arrays.fill(indegree, -1);
        for (String w : words) {
            for (int i = 0; i < w.length(); i++) {
                indegree[w.charAt(i) - 'a'] = 0;
            }
        }

        // 'a' -> 0
        // 'b' -> 1
        // 'z' -> 25
        // x -> x - 'a'
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0, j, len; i < words.length - 1; i++) {
            String cur = words[i];
            String next = words[i + 1];
            j = 0;
            len = Math.min(cur.length(), next.length());
            for (; j < len; ++j) {
                if (cur.charAt(j) != next.charAt(j)) {
                    graph.get(cur.charAt(j) - 'a').add(next.charAt(j) - 'a');
                    indegree[next.charAt(j) - 'a']++;
                    break;
                }
            }

            if (j < cur.length() && j == next.length()) {
                return "";
            }
        }

        PriorityQueue<Integer> heap = new PriorityQueue<>();
        int kinds = 0;
        for (int i = 0; i < 26; i++) {
            if (indegree[i] != -1) {
                ++kinds;
            }
            if (indegree[i] == 0) {
                heap.add(i);
            }
        }

        StringBuilder ans = new StringBuilder();
        while (!heap.isEmpty()) {
            int cur = heap.poll();
            ans.append((char) (cur + 'a'));
            for (int next : graph.get(cur)) {
                if (--indegree[next] == 0) {
                    heap.add(next);
                }
            }
        }

        return ans.length() == kinds ? ans.toString() : "";
    }
}